src/DataTypes/DataTypeMap.h

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include <DataTypes/DataTypeWithSimpleSerialization.h>
+
+
+namespace DB
+{
+
+/** Map data type.
+  *
+  * Map's key and value only have types.
+  * If only one type is set, then key's type is "String" in default.
+  */
+class DataTypeMap final : public DataTypeWithSimpleSerialization
+{
+private:
+    DataTypePtr key_type;
+    DataTypePtr value_type;
+    DataTypePtr nested;
```

Comment:
A (redundant) comment would also be helpful here, e.g. the nested type is an `Array(Tuple(key, value))`.